File: nippon/react-native-indicai/ios/IndicAIModule.swift
import Foundation
import React
import AVFoundation

@objc(IndicAIModule)
class IndicAIModule: RCTEventEmitter {

    private let sdk = IndicAISDK.shared
    private var audioEngine: AVAudioEngine?
    private var recordedPCMData = Data()
    @objc private var isRecording = false

    override static func requiresMainQueueSetup() -> Bool { false }
    override func supportedEvents() -> [String]! { ["IndicAI_Progress"] }

    // Required for RCTEventEmitter
    @objc override func addListener(_ eventName: String) {}
    @objc override func removeListeners(_ count: Double) {}

    @objc func configure(_ manifestUrl: String?, s3BaseUrl: String?) {
        let cfg = SDKConfig(
            manifestURL: URL(string: manifestUrl ?? IndicAISDK.defaultManifestUrl)!,
            s3BaseURL:   URL(string: s3BaseUrl ?? IndicAISDK.defaultS3BaseUrl)!
        )
        Task { await sdk.configure(cfg) }
    }

    @objc func initialize(
        _ languageCode: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                try await sdk.initialize(
                    language: languageCode,
                    onProgress: { [weak self] p in
                        self?.sendEvent(withName: "IndicAI_Progress", body: [
                            "model":      p.modelName,
                            "percent":    p.percent,
                            "downloaded": p.downloaded,
                            "total":      p.total,
                        ])
                    }
                )
                resolver(nil)
            } catch {
                rejecter("INDICAI_INIT_ERROR", error.localizedDescription, error)
            }
        }
    }

    @objc func transcribe(
        _ base64Audio: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard let pcmData = Data(base64Encoded: base64Audio) else {
            rejecter("INDICAI_TRANSCRIBE_ERROR", "Invalid base64 audio data", nil)
            return
        }
        do {
            resolver(try sdk.transcribe(pcmData: pcmData))
        } catch {
            rejecter("INDICAI_TRANSCRIBE_ERROR", error.localizedDescription, error)
        }
    }

    @objc func translateToEnglish(
        _ text: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        do {
            resolver(try sdk.translateToEnglish(text))
        } catch {
            rejecter("INDICAI_TRANSLATE_ERROR", error.localizedDescription, error)
        }
    }

    @objc func translateToIndic(
        _ text: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        do {
            resolver(try sdk.translateToIndic(text))
        } catch {
            rejecter("INDICAI_TRANSLATE_ERROR", error.localizedDescription, error)
        }
    }

    /// Returns the absolute path to a WAV file written in the tmp directory.
    @objc func synthesize(
        _ text: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                let waveform = try await sdk.synthesize(text)
                let path = NSTemporaryDirectory() + "indicai_tts.wav"
                try writeWav(samples: waveform, to: path)
                resolver(path)
            } catch {
                rejecter("INDICAI_TTS_ERROR", error.localizedDescription, error)
            }
        }
    }

    /// Returns { indicText, audioPath } — mirrors Android's respondWithSpeech.
    @objc func respondWithSpeech(
        _ englishText: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                let result = try await sdk.respondWithSpeech(englishText: englishText)
                let path = NSTemporaryDirectory() + "indicai_tts.wav"
                try writeWav(samples: result.audio, to: path)
                resolver(["indicText": result.indicText, "audioPath": path])
            } catch {
                rejecter("INDICAI_RESPOND_ERROR", error.localizedDescription, error)
            }
        }
    }

    /// Classify English text → intent label + confidence. Mirrors Android classifyIntent.
    @objc func classifyIntent(
        _ text: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        do {
            guard let result = try sdk.classifyIntent(text) else {
                rejecter("INDICAI_INTENT_ERROR", "Intent model not loaded", nil)
                return
            }
            resolver(["intent": result.intent, "confidence": result.confidence] as [String: Any])
        } catch {
            rejecter("INDICAI_INTENT_ERROR", error.localizedDescription, error)
        }
    }

    @objc func getSupportedLanguages(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        resolver(Array(LanguageRegistry.supported.keys))
    }

    @objc func isTtsCached(
        _ languageCode: String,
        resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        resolver(sdk.isTtsCached(for: languageCode))
    }

    @objc func release() {
        stopAndReleaseRecorder()
    }

    // ── Native microphone recording (raw PCM Int16 16 kHz mono → base64) ────────

    @objc func startRecording(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard !isRecording else { resolver(nil); return }

        let session = AVAudioSession.sharedInstance()
        do {
            try session.setCategory(.record, mode: .measurement, options: .duckOthers)
            try session.setActive(true, options: .notifyOthersOnDeactivation)

            let engine = AVAudioEngine()
            audioEngine = engine
            let inputNode = engine.inputNode
            // Request 16 kHz mono Int16 — matches what Whisper expects
            let fmt = AVAudioFormat(
                commonFormat: .pcmFormatInt16,
                sampleRate: 16000,
                channels: 1,
                interleaved: true
            )!

            recordedPCMData = Data()
            isRecording = true

            inputNode.installTap(onBus: 0, bufferSize: 1600, format: fmt) { [weak self] buf, _ in
                guard let self = self, self.isRecording else { return }
                let ptr = buf.int16ChannelData![0]
                let bytes = Data(bytes: ptr, count: Int(buf.frameLength) * 2)
                self.recordedPCMData.append(bytes)
            }

            try engine.start()
            resolver(nil)
        } catch {
            rejecter("INDICAI_RECORD_ERROR", error.localizedDescription, error)
        }
    }

    @objc func stopRecording(
        _ resolver: @escaping RCTPromiseResolveBlock,
        rejecter: @escaping RCTPromiseRejectBlock
    ) {
        guard isRecording else { resolver(""); return }
        stopAndReleaseRecorder()
        resolver(recordedPCMData.base64EncodedString())
        recordedPCMData = Data()
    }

    // ── Helpers ──────────────────────────────────────────────────────────────────

    private func stopAndReleaseRecorder() {
        isRecording = false
        audioEngine?.inputNode.removeTap(onBus: 0)
        audioEngine?.stop()
        audioEngine = nil
    }

    /// Write Float32 PCM samples as a 16-bit PCM WAV file.
    private func writeWav(samples: [Float], to path: String) throws {
        let sampleRate: Int32 = 16000
        let dataLen = Int32(samples.count * 2)
        var buf = Data()

        func appendLE<T: FixedWidthInteger>(_ v: T) {
            var le = v.littleEndian
            withUnsafeBytes(of: &le) { buf.append(contentsOf: $0) }
        }

        buf.append(contentsOf: "RIFF".utf8); appendLE(36 + dataLen)
        buf.append(contentsOf: "WAVE".utf8)
        buf.append(contentsOf: "fmt ".utf8); appendLE(Int32(16))
        appendLE(Int16(1));  appendLE(Int16(1))   // PCM, mono
        appendLE(sampleRate); appendLE(sampleRate * 2)
        appendLE(Int16(2));  appendLE(Int16(16))   // block align, bits
        buf.append(contentsOf: "data".utf8); appendLE(dataLen)

        for s in samples {
            appendLE(Int16(max(-32768, min(32767, Int(s * 32767)))))
        }

        try buf.write(to: URL(fileURLWithPath: path))
    }
}

File: nippon/react-native-indicai/android/src/main/java/com/indicai/rn/IndicAIModule.kt
package com.indicai.rn

import android.util.Base64
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.indicai.sdk.IndicAISDK
import com.indicai.sdk.DownloadProgress
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

class IndicAIModule(private val reactContext: ReactApplicationContext)
    : ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "IndicAIModule"

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val sdk = IndicAISDK.getInstance(reactContext)
    private var audioRecord: AudioRecord? = null
    private var recordingJob: Job? = null
    private var recordingBuffer: ByteArrayOutputStream? = null
    @Volatile private var isRecording = false

    @ReactMethod
    fun configure(manifestUrl: String?, s3BaseUrl: String?) {
        val config = IndicAISDK.SDKConfig(
            manifestUrl = manifestUrl ?: IndicAISDK.DEFAULT_MANIFEST_URL,
            s3BaseUrl = s3BaseUrl ?: IndicAISDK.DEFAULT_S3_BASE_URL,
        )
        IndicAISDK.configure(reactContext, config)
    }

    @ReactMethod
    fun initialize(languageCode: String, promise: Promise) {
        sdk.init(
            languageCode = languageCode,
            onProgress = { p: DownloadProgress -> emitProgress(p) },
            onReady = { promise.resolve(null) },
            onError = { e -> promise.reject("INDICAI_INIT_ERROR", e.message, e) },
        )
    }

    @ReactMethod
    fun transcribe(base64Audio: String, promise: Promise) {
        scope.launch {
            try {
                val pcmBytes = Base64.decode(base64Audio, Base64.NO_WRAP)
                val result = sdk.transcribe(pcmBytes)
                promise.resolve(result)
            } catch (e: Exception) {
                promise.reject("INDICAI_TRANSCRIBE_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun translateToEnglish(text: String, promise: Promise) {
        scope.launch {
            try {
                promise.resolve(sdk.translateToEnglish(text))
            } catch (e: Exception) {
                promise.reject("INDICAI_TRANSLATE_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun translateToIndic(text: String, promise: Promise) {
        scope.launch {
            try {
                promise.resolve(sdk.translateToIndic(text))
            } catch (e: Exception) {
                promise.reject("INDICAI_TRANSLATE_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun synthesize(text: String, promise: Promise) {
        sdk.synthesize(
            text = text,
            onProgress = { p -> emitProgress(p) },
            onResult = { pcmFloat ->
                val outFile = File(reactContext.cacheDir, "indicai_tts.wav")
                writeWav(pcmFloat, outFile)
                promise.resolve(outFile.absolutePath)
            },
            onError = { e -> promise.reject("INDICAI_TTS_ERROR", e.message) },
        )
    }

    @ReactMethod
    fun respondWithSpeech(englishText: String, promise: Promise) {
        sdk.respondWithSpeech(
            englishText = englishText,
            onProgress = { p -> emitProgress(p) },
            onResult = { indicText, audio ->
                val outFile = File(reactContext.cacheDir, "indicai_tts.wav")
                writeWav(audio, outFile)
                val result = Arguments.createMap().apply {
                    putString("indicText", indicText)
                    putString("audioPath", outFile.absolutePath)
                }
                promise.resolve(result)
            },
            onError = { e -> promise.reject("INDICAI_RESPOND_ERROR", e.message) },
        )
    }

    /** Write Float32 PCM samples as a 16-bit PCM WAV file. */
    private fun writeWav(samples: FloatArray, file: File) {
        val dataBytes = samples.size * 2
        val sampleRate = 16000
        val buf = ByteBuffer.allocate(44 + dataBytes).order(ByteOrder.LITTLE_ENDIAN)
        buf.put("RIFF".toByteArray(Charsets.US_ASCII))
        buf.putInt(36 + dataBytes)
        buf.put("WAVE".toByteArray(Charsets.US_ASCII))
        buf.put("fmt ".toByteArray(Charsets.US_ASCII))
        buf.putInt(16)
        buf.putShort(1)
        buf.putShort(1)
        buf.putInt(sampleRate)
        buf.putInt(sampleRate * 2)
        buf.putShort(2)
        buf.putShort(16)
        buf.put("data".toByteArray(Charsets.US_ASCII))
        buf.putInt(dataBytes)
        for (s in samples) {
            buf.putShort((s.coerceIn(-1f, 1f) * 32767f).toInt().toShort())
        }
        file.writeBytes(buf.array())
    }

    @ReactMethod
    fun classifyIntent(text: String, promise: Promise) {
        scope.launch {
            try {
                val result = sdk.classifyIntent(text)
                if (result == null) {
                    promise.reject("INDICAI_INTENT_ERROR", "Intent model not loaded")
                } else {
                    val map = Arguments.createMap().apply {
                        putString("intent", result.intent)
                        putDouble("confidence", result.confidence.toDouble())
                    }
                    promise.resolve(map)
                }
            } catch (e: Exception) {
                promise.reject("INDICAI_INTENT_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod
    fun getSupportedLanguages(promise: Promise) {
        val arr = Arguments.createArray()
        sdk.getSupportedLanguages().forEach { arr.pushString(it) }
        promise.resolve(arr)
    }

    @ReactMethod
    fun isTtsCached(languageCode: String, promise: Promise) {
        promise.resolve(sdk.isTtsCachedFor(languageCode))
    }

    @ReactMethod
    fun release() {
        stopAndReleaseRecorder()
        sdk.release()
    }

    @ReactMethod
    fun startRecording(promise: Promise) {
        try {
            if (isRecording) {
                promise.resolve(null)
                return
            }
            val sampleRate = 16000
            val minBuffer = AudioRecord.getMinBufferSize(
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT
            )
            if (minBuffer <= 0) {
                promise.reject("INDICAI_RECORD_ERROR", "Invalid recorder buffer size")
                return
            }
            val recorder = AudioRecord(
                MediaRecorder.AudioSource.MIC,
                sampleRate,
                AudioFormat.CHANNEL_IN_MONO,
                AudioFormat.ENCODING_PCM_16BIT,
                minBuffer * 2
            )
            if (recorder.state != AudioRecord.STATE_INITIALIZED) {
                recorder.release()
                promise.reject("INDICAI_RECORD_ERROR", "AudioRecord init failed")
                return
            }

            audioRecord = recorder
            recordingBuffer = ByteArrayOutputStream()
            isRecording = true
            recorder.startRecording()

            recordingJob = scope.launch {
                val chunk = ByteArray(minBuffer)
                while (isRecording) {
                    val n = recorder.read(chunk, 0, chunk.size)
                    if (n > 0) {
                        recordingBuffer?.write(chunk, 0, n)
                    }
                }
            }
            promise.resolve(null)
        } catch (e: Exception) {
            stopAndReleaseRecorder()
            promise.reject("INDICAI_RECORD_ERROR", e.message, e)
        }
    }

    @ReactMethod
    fun stopRecording(promise: Promise) {
        scope.launch {
            try {
                if (!isRecording) {
                    promise.resolve("")
                    return@launch
                }
                isRecording = false
                recordingJob?.join()
                val recorder = audioRecord
                try { recorder?.stop() } catch (_: Exception) {}
                recorder?.release()
                audioRecord = null
                recordingJob = null

                val pcm = recordingBuffer?.toByteArray() ?: ByteArray(0)
                recordingBuffer = null
                val b64 = Base64.encodeToString(pcm, Base64.NO_WRAP)
                promise.resolve(b64)
            } catch (e: Exception) {
                stopAndReleaseRecorder()
                promise.reject("INDICAI_RECORD_ERROR", e.message, e)
            }
        }
    }

    @ReactMethod fun addListener(eventName: String) {}
    @ReactMethod fun removeListeners(count: Int) {}

    private fun emitProgress(p: DownloadProgress) {
        val map = Arguments.createMap().apply {
            putString("model", p.modelName)
            putInt("percent", p.percent)
            putDouble("downloaded", p.downloadedBytes.toDouble())
            putDouble("total", p.totalBytes.toDouble())
        }
        reactContext
            .getJSModule(RCTDeviceEventEmitter::class.java)
            .emit("IndicAI_Progress", map)
    }

    private fun stopAndReleaseRecorder() {
        isRecording = false
        try { audioRecord?.stop() } catch (_: Exception) {}
        try { audioRecord?.release() } catch (_: Exception) {}
        audioRecord = null
        recordingJob?.cancel()
        recordingJob = null
        recordingBuffer = null
    }
}
